tools/depmat.py

```diff
@@ -0,0 +1,181 @@
+			matrix_cell_offset = matrix_row_offset + 3*matrix_x
+			addend_cell_offset = addend_row_offset + 3*addend_x
+			matrix.data[matrix_cell_offset+0] += addend.data[addend_cell_offset+0]
+			matrix.data[matrix_cell_offset+1] += addend.data[addend_cell_offset+1]
+			matrix.data[matrix_cell_offset+2] += addend.data[addend_cell_offset+2]
+
+def change_mapping(matrix, new_mapping):
+	return DepMat(matrix.size, [new_mapping[old] for old in matrix.mapping], matrix.data)
+
+def empty(size):
+	return DepMat(size, None, [0] * (size*size*3))
+
+####
+
+def html(event):
+	# TODO adapt from `viewers.depmat`
+	print(f'<!doctype html><html><head><title>{event.event_id}.depmat</title></head><body>TODO</body></html>')
+
+def csv(event):
+	# XXX define format
+	print('0,0,0')
+
+def aggregate(events):
+	aggregated_ps, mappings = phasestack.aggregate(events)
+	aggregated_matrix = empty(aggregated_ps.size)
+	for i, event in enumerate(events):
+		matrix = event.get('depmat', load)
+		add_to(aggregated_matrix, change_mapping(matrix, mappings[i]))
+	return aggregated_matrix, aggregated_ps
+
+def diff(from_, to):
+	# TODO do something useful (currently just lame matrix subtraction)
+	from_depmat = from_.get('depmat', load)
+	from_ps = from_.get('phasestack', phasestack.load)
+	neg_from_depmat = DepMat(from_depmat.size, from_depmat.mapping, [-value for value in from_depmat.data])
+	neg_from = Event(depmat=neg_from_depmat, phasestack=from_ps)
+	return aggregate([neg_from, to])
+
+def expand(original_event, new_ps_event):
+	original_matrix = original_event.get('depmat', load)
+	original_ps = original_event.get('phasestack', phasestack.load)
+	new_ps = new_ps_event.get('phasestack', phasestack.load)
+	mapping = phasestack.get_mapping(original_ps, new_ps)
+	new_matrix = change_mapping(original_matrix, mapping)
+	return new_matrix, new_ps
+
+####
+
+def main():
+	try:
+		[_script_name, function, *args] = sys.argv
+	except ValueError:
+		raise UserError('Usage: depmat <FUNCTION> [<ARGS> ...]')
+	function = function.lstrip('-')
+	opts = None
+	
+	def scan_options(longopts_str):
+		nonlocal opts, args
+		opts_list, args_list = gnu_getopt(args, '', longopts_str.split())
+		opts_dict = dict(opts_list)
+		if len(opts_dict) != len(opts_list):
+			raise UserError('Duplicate option!')
+		opts = opts_dict
+		args = args_list
+	
+	if function in ('help', 'usage'):
+		print('Usage:')
+		print('  depmat help')
+		print('  depmat html [--tty] <EVENT>')
+		print('  depmat csv [--tty] <EVENT>')
+		print('  depmat aggregate <EVENT LIST ...> [--as <TARGET>]')
+		print('  depmat diff <FROM_EVENT> <TO_EVENT> [--as <TARGET>]')
+		print('  depmat expand <ORIGINAL> <NEW_PHASESTACK> [--as <TARGET>]')
+	
+	elif function == 'html':
+		scan_options('tty')
+		tty = '--tty' in opts
+		if not tty and sys.stdout.isatty():
+			raise UserError('Refusing to print to terminal. Redirect output or use --tty')
+		(event,) = n_args_to_n_events(1, args, 'Usage: depmat html [--tty] <EVENT>')
```

Comment:
There is no need to use brackets when using tuple unpacking. I haven't actually seen it before. And in this specific case, it's not clear, whether:

```
event, _* = ...
```

would be any better than what you wrote or just simply this:

```
event = ...[0]
```

I think I prefer the last method...